lib/src/Base/Algo/ClenshawCurtis.cxx

```diff
@@ -0,0 +1,212 @@
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/ClenshawCurtis.hxx"
+#include "openturns/Tuples.hxx"
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/Lapack.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+/**
+ * @class ClenshawCurtis
+ */
+
+CLASSNAMEINIT(ClenshawCurtis)
+
+static const Factory<ClenshawCurtis> Factory_ClenshawCurtis;
+
+/* Constructor without parameters */
+ClenshawCurtis::ClenshawCurtis(const UnsignedInteger dimension)
+  : IntegrationAlgorithmImplementation()
+  , discretization_(Indices(dimension, ResourceMap::GetAsUnsignedInteger("ClenshawCurtis-DefaultMarginalIntegrationPointsNumber")))
+  , nodes_(0, 0)
+  , weights_(0)
+{
+  // Generate nodes and weights
+  generateNodesAndWeights();
+}
+
+/* Parameters constructor */
+ClenshawCurtis::ClenshawCurtis(const Indices & discretization)
+  : IntegrationAlgorithmImplementation()
+  , discretization_(discretization)
+  , nodes_(0, 0)
+  , weights_(0)
+{
+  // Generate nodes and weights
+  generateNodesAndWeights();
+}
+
+/* Virtual constructor */
+ClenshawCurtis * ClenshawCurtis::clone() const
+{
+  return new ClenshawCurtis(*this);
+}
+
+/* Compute an approximation of \int_a^b f(x_1,\dots,x_n)dx_1\dotsdx_n, where [a,b] is an n-D interval.
+ */
+Point ClenshawCurtis::integrate(const Function & function,
+                               const Interval & interval) const
+{
+  Sample adaptedNodes;
+  return integrateWithNodes(function, interval, adaptedNodes);
+}
+
+Point ClenshawCurtis::integrateWithNodes(const Function & function,
+                                        const Interval & interval,
+                                        Sample & adaptedNodes) const
+{
+  const UnsignedInteger inputDimension = discretization_.getSize();
+  if (interval.getDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: expected an interval of dimension=" << inputDimension << ", got dimension=" << interval.getDimension();
+  if (function.getInputDimension() != inputDimension) throw InvalidArgumentException(HERE) << "Error: expected a function of input dimension=" << inputDimension << ", got input dimension=" << function.getInputDimension();
+  // Compute the volume of the interval
+  const Scalar volume = interval.getVolume();
+  Point integral(function.getOutputDimension());
+  if (volume == 0.0) return integral;
+  // Adapt the nodes to the bounds of the interval
+  adaptedNodes = nodes_ * (interval.getUpperBound() - interval.getLowerBound())/2. + interval.getLowerBound()+(interval.getUpperBound() - interval.getLowerBound())/2.;
+  // Compute the function over the adapted nodes
+  const Sample values(function(adaptedNodes));
+  // Compute the integral
+  for (UnsignedInteger i = 0; i < values.getSize(); ++i)
+    integral += values[i] * weights_[i];
+  return std::pow(0.5,inputDimension)*integral * volume;
```

Comment:
unfortunately this line does not build on windows compiler, it should work with:
`return std::pow(0.5, (Scalar)inputDimension)*integral * volume;`